ClosedXML/Excel/Style/XLFill.cs

```diff
@@ -5,142 +5,163 @@ namespace ClosedXML.Excel
+            {
+                key = (defaultFill as XLFill).Key;
+            }
+            else
+            {
+                key = new XLFillKey
+                {
+                    PatternType = defaultFill.PatternType,
+                    BackgroundColor = defaultFill.BackgroundColor.Key,
+                    PatternColor = defaultFill.PatternColor.Key
+                };
+            }
+            return key;
         }
 
-        #endregion IXLFill Members
+        #endregion static members
+
+        #region Properties
+
+        private readonly XLStyle _style;
+
+        private XLFillValue _value;
 
-        private void SetStyleChanged()
+        internal XLFillKey Key
         {
-            if (_container != null) _container.StyleChanged = true;
+            get { return _value.Key; }
+            private set { _value = XLFillValue.FromKey(value); }
         }
 
-        public override bool Equals(object obj)
+        #endregion Properties
+
+        #region Constructors
+
+        /// <summary>
+        /// Create an instance of XLFill initializing it with the specified value.
+        /// </summary>
+        /// <param name="style">Style to attach the new instance to.</param>
+        /// <param name="value">Style value to use.</param>
+        public XLFill(XLStyle style, XLFillValue value)
         {
-            return Equals((XLFill)obj);
+            _style = style ?? XLStyle.CreateEmptyStyle();
+            _value = value;
         }
 
-        public override int GetHashCode()
+        public XLFill(XLStyle style, XLFillKey key) : this(style, XLFillValue.FromKey(key))
         {
-            return BackgroundColor.GetHashCode()
-                   ^ (Int32)PatternType
-                   ^ PatternColor.GetHashCode();
         }
 
-        #region Properties
+        public XLFill(XLStyle style = null, IXLFill d = null) : this(style, GenerateKey(d))
+        {
+        }
+
+        #endregion Constructors
+
+        private void Modify(Func<XLFillKey, XLFillKey> modification)
+        {
+            Key = modification(Key);
 
-        private XLColor _backgroundColor;
-        private XLColor _patternColor;
-        private XLFillPatternValues _patternType;
+            _style.Modify(styleKey =>
+            {
+                var fill = styleKey.Fill;
+                styleKey.Fill = modification(fill);
+                return styleKey;
+            });
+        }
+
+        #region IXLFill Members
 
         public XLColor BackgroundColor
         {
-            get { return _backgroundColor; }
+            get { return XLColor.FromKey(Key.BackgroundColor); }
             set
             {
-                SetStyleChanged();
-                if (_container != null && !_container.UpdatingStyle)
-                    _container.Styles.ForEach(s => s.Fill.BackgroundColor = value);
-                else
+                if (value == null)
+                    throw new ArgumentNullException("Color cannot be null");
```

Comment:
Please change all these to e.g. `throw new ArgumentNullException(nameof(value), "Color cannot be null")` . The overload you're currently using is `ArgumentNullException(string paramName)`, which isn't supposed to include a message, only paramname.

Reply:
Done